aws/endpoints/v3model_codegen.go

```diff
@@ -0,0 +1,284 @@
+// +build codegen
+
+package endpoints
+
+import (
+	"fmt"
+	"io"
+	"reflect"
+	"strings"
+	"text/template"
+)
+
+// A CodeGenOptions are the options for code generating the endpoints into
+// Go code from the endpoints model definition.
+type CodeGenOptions struct {
+	// Options for how the model will be decoded.
+	DecodeModelOptions DecodeModelOptions
+}
+
+// Set combines all of the option functions together
+func (d *CodeGenOptions) Set(optFns ...func(*CodeGenOptions)) {
+	for _, fn := range optFns {
+		fn(d)
+	}
+}
+
+// CodeGenModel given a endpoints model file will decode it and attempt to
+// generate Go code from the model definition. Error will be returned if
+// the code is unable to be generated, or decoded.
+func CodeGenModel(modelFile io.Reader, outFile io.Writer, optFns ...func(*CodeGenOptions)) error {
+	var opts CodeGenOptions
+	opts.Set(optFns...)
+
+	resolver, err := DecodeModel(modelFile, func(d *DecodeModelOptions) {
+		*d = opts.DecodeModelOptions
+	})
+	if err != nil {
+		return err
+	}
+
+	tmpl := template.Must(template.New("tmpl").Funcs(funcMap).Parse(v3Tmpl))
+	if err := tmpl.ExecuteTemplate(outFile, "defaults", resolver); err != nil {
+		return fmt.Errorf("failed to execute template, %v", err)
+	}
+
+	return nil
+}
+
+func camelCase(v string) string {
+	return strings.Replace(strings.Title(v), "-", "", -1)
+}
+
+func quoteString(v string) string {
+	return fmt.Sprintf("%q", v)
+}
+
+func regionConstName(p, r string) string {
+	return camelCase(p) + camelCase(r)
+}
+
+func partitionGetter(id string) string {
+	return fmt.Sprintf("%sPartition", camelCase(id))
+}
+
+func partitionVarName(id string) string {
+	return fmt.Sprintf("%sPartition", strings.ToLower(camelCase(id)))
+}
+
+func listPartitionNames(ps partitions) string {
+	names := []string{}
+	for i, p := range ps {
+		if i == len(ps)-1 {
+			names = append(names, "and "+p.Name)
+		} else {
+			names = append(names, p.Name)
+		}
+	}
+	return strings.Join(names, ", ")
```

Comment:
if `len(ps) == 2` wouldn't the resulting string be `foo, and bar` when it should be `foo and bar`?